File: rust/tests/power/ina3221_test/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::power::{Ina3221Full, CF1, CF2, CF3, WF1, WF2, WF3, PVF, TCF, CVRF};

macro_rules! check_true {
    ($cond:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $cond {
            println!("PASS {}", $label);
            $passed += 1;
        } else {
            println!("FAIL {}", $label);
            $failed += 1;
        }
    };
}

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x40);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut chip = Ina3221Full::new(dev, addr, 0.1).expect("init INA3221");

    let mut passed = 0i32;
    let mut failed = 0i32;

    check_true!(
        chip.manufacturer_id().unwrap() == 0x5449,
        "manufacturer_id",
        passed,
        failed
    );
    check_true!(
        chip.die_id().unwrap() == 0x3220,
        "die_id",
        passed,
        failed
    );

    for ch in 1..=3 {
        let label = format!("ch{}_voltage_non_negative", ch);
        check_true!(chip.voltage(ch).unwrap() >= 0.0, label, passed, failed);

        let label = format!("ch{}_shunt_voltage_finite", ch);
        check_true!(chip.shunt_voltage(ch).unwrap().abs() < 1.0, label, passed, failed);

        let label = format!("ch{}_current_finite", ch);
        check_true!(chip.current(ch).unwrap().abs() < 100.0, label, passed, failed);

        let label = format!("ch{}_power_non_negative", ch);
        check_true!(chip.power(ch).unwrap() >= 0.0, label, passed, failed);
    }

    check_true!(
        chip.conversion_ready().is_ok(),
        "conversion_ready_ok",
        passed,
        failed
    );

    chip.configure(3, 4, 4, 7).unwrap();
    check_true!(
        chip.manufacturer_id().unwrap() == 0x5449,
        "configure_mfr_id_still_valid",
        passed,
        failed
    );

    chip.set_critical_alert(1, 0.1, false).unwrap();
    chip.set_warning_alert(2, 0.05, false).unwrap();
    let flags = chip.alert_flags().unwrap();
    check_true!(flags >= 0, "alert_flags_readable", passed, failed);

    chip.enable_channel(1, false).unwrap();
    check_true!(
        !chip.channel_enabled(1).unwrap(),
        "channel_1_disabled",
        passed,
        failed
    );
    chip.enable_channel(1, true).unwrap();
    check_true!(
        chip.channel_enabled(1).unwrap(),
        "channel_1_re_enabled",
        passed,
        failed
    );

    chip.set_summation_channels(&[1, 2], 0.2).unwrap();
    let sv_sum = chip.summation_value().unwrap();
    check_true!(sv_sum.abs() < 10.0, "summation_value_finite", passed, failed);

    chip.set_power_valid_limits(5.5, 4.5).unwrap();
    check_true!(
        chip.power_valid().is_ok(),
        "power_valid_readable",
        passed,
        failed
    );

    chip.shutdown().unwrap();
    chip.wake().unwrap();
    check_true!(
        chip.voltage(1).unwrap() >= 0.0,
        "wake_voltage_non_negative",
        passed,
        failed
    );

    chip.reset().unwrap();
    check_true!(
        chip.manufacturer_id().unwrap() == 0x5449,
        "reset_mfr_id_still_valid",
        passed,
        failed
    );

    println!("===DONE: {} passed, {} failed===", passed, failed);
    std::process::exit(if failed == 0 { 0 } else { 1 });
}

File: go/tests/power/ina3221_test/main.go
//go:build linux && !tinygo

// INA3221 hardware test — Linux host.
//
// Reads from /dev/i2c-N and runs the INA3221 check sequence. Prints
// PASS/FAIL per check and ends with the standard ===DONE: ... === line.
// Exits 0 on full pass, 1 on any failure.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/power"
	"github.com/tuhde/Periph/go/periph/transport"
)

func main() {
	bus, err := strconv.Atoi(envOr("I2C_BUS", "1"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_BUS:", err)
		os.Exit(2)
	}
	addr, err := strconv.ParseUint(envOr("I2C_ADDR", "0x40"), 0, 8)
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_ADDR:", err)
		os.Exit(2)
	}

	tr, err := transport.NewI2CTransport(bus, uint8(addr))
	if err != nil {
		fmt.Fprintln(os.Stderr, "transport:", err)
		os.Exit(2)
	}
	defer tr.Close()

	chip, err := power.NewINA3221Full(tr, 0.1)
	if err != nil {
		fmt.Fprintln(os.Stderr, "new:", err)
		os.Exit(2)
	}

	passed, failed := 0, 0
	check := func(label string, cond bool) {
		if cond {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	mfr, err := chip.ManufacturerID()
	check("manufacturer_id", err == nil && mfr == 0x5449)

	die, err := chip.DieID()
	check("die_id", err == nil && die == 0x3220)

	for ch := uint8(1); ch <= 3; ch++ {
		v, err := chip.Voltage(ch)
		label := fmt.Sprintf("ch%d_voltage_non_negative", ch)
		check(label, err == nil && v >= 0.0)

		_, err = chip.ShuntVoltage(ch)
		label = fmt.Sprintf("ch%d_shunt_voltage_finite", ch)
		check(label, err == nil)

		_, err = chip.Current(ch)
		label = fmt.Sprintf("ch%d_current_finite", ch)
		check(label, err == nil)

		p, err := chip.Power(ch)
		label = fmt.Sprintf("ch%d_power_non_negative", ch)
		check(label, err == nil && p >= 0.0)
	}

	_, err = chip.ConversionReady()
	check("conversion_ready_ok", err == nil)

	if err := chip.Configure(3, 4, 4, 7); err != nil {
		fmt.Fprintln(os.Stderr, "configure:", err)
		os.Exit(2)
	}
	if mfr, err := chip.ManufacturerID(); err == nil && mfr == 0x5449 {
		check("configure_mfr_id_still_valid", true)
	} else {
		check("configure_mfr_id_still_valid", false)
	}

	if err := chip.SetCriticalAlert(1, 0.1, false); err != nil {
		fmt.Fprintln(os.Stderr, "set_critical_alert:", err)
		os.Exit(2)
	}
	if err := chip.SetWarningAlert(2, 0.05, false); err != nil {
		fmt.Fprintln(os.Stderr, "set_warning_alert:", err)
		os.Exit(2)
	}
	if _, err := chip.AlertFlags(); err == nil {
		check("alert_flags_readable", true)
	} else {
		check("alert_flags_readable", false)
	}

	if err := chip.EnableChannel(1, false); err != nil {
		fmt.Fprintln(os.Stderr, "disable_channel:", err)
		os.Exit(2)
	}
	if enabled, err := chip.ChannelEnabled(1); err == nil && !enabled {
		check("channel_1_disabled", true)
	} else {
		check("channel_1_disabled", false)
	}
	if err := chip.EnableChannel(1, true); err != nil {
		fmt.Fprintln(os.Stderr, "enable_channel:", err)
		os.Exit(2)
	}
	if enabled, err := chip.ChannelEnabled(1); err == nil && enabled {
		check("channel_1_re_enabled", true)
	} else {
		check("channel_1_re_enabled", false)
	}

	if err := chip.SetSummationChannels([]uint8{1, 2}, 0.2); err != nil {
		fmt.Fprintln(os.Stderr, "set_summation_channels:", err)
		os.Exit(2)
	}
	if _, err := chip.SummationValue(); err == nil {
		check("summation_value_readable", true)
	} else {
		check("summation_value_readable", false)
	}

	if err := chip.SetPowerValidLimits(5.5, 4.5); err != nil {
		fmt.Fprintln(os.Stderr, "set_power_valid_limits:", err)
		os.Exit(2)
	}
	if _, err := chip.PowerValid(); err == nil {
		check("power_valid_readable", true)
	} else {
		check("power_valid_readable", false)
	}

	if err := chip.Shutdown(); err != nil {
		fmt.Fprintln(os.Stderr, "shutdown:", err)
		os.Exit(2)
	}
	if err := chip.Wake(); err != nil {
		fmt.Fprintln(os.Stderr, "wake:", err)
		os.Exit(2)
	}
	if v, err := chip.Voltage(1); err == nil && v >= 0.0 {
		check("wake_voltage_non_negative", true)
	} else {
		check("wake_voltage_non_negative", false)
	}

	if err := chip.Reset(); err != nil {
		fmt.Fprintln(os.Stderr, "reset:", err)
		os.Exit(2)
	}
	if mfr, err := chip.ManufacturerID(); err == nil && mfr == 0x5449 {
		check("reset_mfr_id_still_valid", true)
	} else {
		check("reset_mfr_id_still_valid", false)
	}

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}
